### PR TITLE
Add memory swap @open sesame 10/13 14:17

### PR DESCRIPTION
**Pull request by jihochu**
Initial implementaion of cache pool.
It inherites from memory pool to utilize optimized memory information.
And, swap device class is introduce. It operates with cache memory and keeps data permantly in the device.
The storage size is fixed when the device opens, sparse file is created with garbage data. And, it's removed when the device close, 

Signed-off-by: Jiho Chu <jiho.chu@samsung.com>

Issue: #1966

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

**Comment by lhs8928** [comment]
Please add cache_pool.cpp and swap_device.cpp at jni/Android.mk to resolve android ci issue

File: nntrainer/tensor/cache_pool.h
Comment (jihochu) [reply nntrainer/tensor/cache_pool.h:19]: Thanks. But, this PR includes many files, and dependent modules. I organized commits for better reviewing.
I think you can understand this.

File: nntrainer/tensor/cache_pool.h
Comment (jihochu) [reply nntrainer/tensor/cache_pool.h:73]: ditto.

File: nntrainer/tensor/cache_pool.h
Comment (jihochu) [reply nntrainer/tensor/cache_pool.h:82]: It changed to /tmp.
Configurable swap path is preparing, but I think the default seems to be necessary anyway.

File: nntrainer/tensor/cache_pool.h
Comment (jihochu) [reply nntrainer/tensor/cache_pool.h:111]: It fixed.

File: nntrainer/tensor/swap_device.cpp
Comment (jihochu) [reply nntrainer/tensor/swap_device.cpp:28]: I think `ret `is better name for `write` and `seek` return values.

File: nntrainer/tensor/swap_device.cpp
Comment (jihochu) [reply nntrainer/tensor/swap_device.cpp:45]: The APIs are to give oppertunity when it start/end device operation. It's helpful to detect unexptected get/put buffer ops.

File: nntrainer/tensor/swap_device.cpp
Comment (jihochu) [reply nntrainer/tensor/swap_device.cpp:66]: thanks. it modiied.

File: nntrainer/tensor/swap_device.cpp
Comment (jihochu) [reply nntrainer/tensor/swap_device.cpp:87]: thanks. you captured bug. ^^

File: nntrainer/tensor/swap_device.cpp
Comment (jihochu) [reply nntrainer/tensor/swap_device.cpp:94]: It modifed.

File: nntrainer/tensor/swap_device.h
Comment (jihochu) [reply nntrainer/tensor/swap_device.h:10]: thanks.

File: nntrainer/tensor/swap_device.h
Comment (jihochu) [reply nntrainer/tensor/swap_device.h:28]: It's intended, but I added more info.

File: nntrainer/tensor/swap_device.h
Comment (jihochu) [reply nntrainer/tensor/swap_device.h:42]: It calls default contructor and it sets empty container.

File: nntrainer/tensor/swap_device.h
Comment (jihochu) [reply nntrainer/tensor/swap_device.h:58]: thanks. comments are aded.

File: nntrainer/tensor/tensor.cpp
Comment (jihochu) [reply nntrainer/tensor/tensor.cpp:164]: please check Deleter.
https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr

**Review by jihochu — COMMENTED** [review]
Thanks for your review, and sorry for late reply.
All modification is updated, but the revision is being ready.

**Comment by jihochu** [comment]
It depends on #1987 . Please ignore build error before merging it.

File: nntrainer/tensor/cache_pool.h
Comment (myungjoo) [reply nntrainer/tensor/cache_pool.h]: You may have such defined in "/ext/nntrainer.ini" with a dynamic default value of https://valadoc.org/gio-2.0/GLib.File.new_tmp.html .

Comment (jihochu) [reply nntrainer/tensor/cache_pool.h]: Thanks. I'll modify this in the next PR.

**Comment by myungjoo** [comment]
Added "RELEASE CRITICAL": As long as this is an experimental feature that is turned off by default and can be turned on with users' explicit configuration, this is not going to affect the behaviors.

For further testing and verification, let's merge this as long as it does not affect the behaviors with the default configuration.

File: nntrainer/tensor/tensor_pool.h
Comment (jijoongmoon) [inline nntrainer/tensor/tensor_pool.h]: I wonder do we have to change this? ```CachePool``` inherit ```MemoryPool``` and does it work if we turn the swap option off?

Comment (jihochu) [reply nntrainer/tensor/tensor_pool.h]: It modified in later commit.
I removed this chages from the commit.

File: nntrainer/tensor/cache_pool.h
Comment (jijoongmoon) [inline nntrainer/tensor/cache_pool.h:96]: It could be considered later PR, but how about enherit the MemoryData for CacheElem, rather than has it.

File: nntrainer/tensor/swap_device.h
Comment (jijoongmoon) [inline nntrainer/tensor/swap_device.h:44]: It might be better set this using property from user in later PR

File: nntrainer/tensor/cache_pool.cpp
Comment (jijoongmoon) [inline nntrainer/tensor/cache_pool.cpp:48]: I think the swap_device is needs to be set for every ```CachePool``` and I think we might needs multiple memory pool in later usage scenario. So.. setting the swap device name should not be set by using just getpid().
It might be considered in later PR though.

File: nntrainer/tensor/swap_device.h
Comment (jijoongmoon) [inline nntrainer/tensor/swap_device.h:110]: I'm not sure but do you think it might be better to use multiple fd w.r.t performance. Cause.. there might be a case which using swap device over 1GB memory. So.. if we knew the offset and get and put using multiple fd, then I think it could be better. Let me know your thought.

File: nntrainer/tensor/cache_pool.h
Comment (jihochu) [reply nntrainer/tensor/cache_pool.h:96]: The chracteristic of MemoryData and CacheElem is little different. I'll investigate how it can be inherited.

File: nntrainer/tensor/swap_device.h
Comment (jihochu) [reply nntrainer/tensor/swap_device.h:44]: OK. I'll add other property.

File: nntrainer/tensor/cache_pool.cpp
Comment (jihochu) [reply nntrainer/tensor/cache_pool.cpp:48]: Multiple memory pool case is not considered. I'll think of it more,
Thanks for your advice.

File: nntrainer/tensor/swap_device.h
Comment (jihochu) [reply nntrainer/tensor/swap_device.h:110]: Performance comparision is needed for it.
I'm preparing IO perf measure, so let's discuss more with the benchmark result.

File: nntrainer/graph/network_graph.cpp
Comment (jihochu) [reply nntrainer/graph/network_graph.cpp:338]: It would be great with disabling memory-swap  while traning job. I'll add this for later PR.
Thanks.

**Review by jijoongmoon — APPROVED** [review]
LGTM!

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Pull request merged**
